### ISSUE TITLE
Menu refreshing does not happen outside of viewport

### ISSUE THREAD

**Issue by SilverGreen93** [koreader/koreader#11316]
* KOReader version: 2023.10 + #11302
* Device: PB743K3

#### Issue
After pushing the #11302 viewport for Pocketbook Color 3, I noticed that when reading reflowable (epub) books, opening and closing the top or bottom menus leave a reminiscence of the black line margin on the area outside the defined viewport.

If I tilt the reader slightly, I can better see the line:
![IMG_20240103_124138](https://github.com/koreader/koreader/assets/24845602/e1a41799-c4ca-4daf-8942-a9a57cdbfb00)

#### Steps to reproduce
1. Open the top or bottom menu.
2. Close the menu.
3. Notice a couple of pixels of the black bar just remaining on the edge of the screen (outside of the defined viewport).

#### Possible solutions
1. Make the menus not draw outside of the viewport.
2. Make the menus redraw the lines (refresh) outside the viewport as well to make the line disappear.

Which solution would be best?

**Comment by Frenzie** [comment]
Assuming your analysis is correct, 1 is obviously correct or at least not wrong and 2 is lunacy. ;-)

Someone will have to check later but I wouldn't necessarily assume PocketBook is quite the same as other devices ftr.

https://github.com/koreader/koreader-base/blob/d83148b27e25b9332678221f6737d66be7e73318/ffi/framebuffer_einkfb.lua#L9-L24

https://github.com/koreader/koreader-base/blob/d83148b27e25b9332678221f6737d66be7e73318/ffi/framebuffer_pocketbook.lua#L12-L17

https://github.com/koreader/koreader-base/blob/d83148b27e25b9332678221f6737d66be7e73318/ffi/framebuffer_pocketbook.lua#L49-L59

**Comment by poire-z** [comment]
I remember that when building UI stuff, it has happened that I made stuff bigger than the screen and have it centered so parts of the bigger stuff is offscreen and not shown.
Ie. for this TouchMenu, I think that for the bottom thick line, I just have it by making the menu a FrameContainer with a 4-sided thick border (we don't have code to draw border on some sides only), and the top, left and right thick borders end up being off-screen, and only the bottom one is visible. (Would have been less easy to add a bottom thick LineWidget.)
I may have used that trick in other places, maybe the Page browser viewfinder extending over the screen bottom.
Also, we may have stuff that overflows the screen and that we've been fine with not caring (ie. the very old Book status widget that may get bigger than the screen/viewport in some conditions).

So, I hope this can be fixed in the refresh code or the painting code (if it's not too expensive checking for overflow for each pixel setting...) - but I guess we would also want to paint that viewport overflow white so you don't see in there old crap from the OS (btw, does the Pocketbook OS ensure that smaller viewport?).
How does nightmode play with that viewport ? Do we also invert the screen outside of that viewport?

**Comment by Frenzie** [comment]
> Do we also invert the screen outside of that viewport?

No (unless it's different for hardware support).

**Comment by NiLuJe** [comment]
3. Fix the viewport to include those pixels ;).

**Comment by Frenzie** [comment]
That's part of what I meant by "at least not wrong". I don't think it entails a promise of not drawing outside, even if I might not necessarily expect it to. I think it only entails a promise that what you want to see will be inside.

**Comment by NiLuJe** [comment]
Blitting will *never* allow blitting outside of the bb constraints, and when a viewport is in place, that's the constraints of said viewport, not of the full bb.

So, if it's painted, it's *inside* the viewport.

There *might* be an off-by-one somewhere in the rotation code (`bb:getPjysicalRect`), or the viewport offset accounting during refreshes (Renamed to `fb:calculateRealCoordinates` in https://github.com/koreader/koreader-base/pull/1718)

**Comment by Frenzie** [comment]
> Blitting will _never_ allow blitting outside of the bb constraints, and when a viewport is in place, that's the constraints of said viewport, not of the full bb.

That's what I would expect, but note that I cleverly wrote "assuming" in my initial reply. ;-)

**Comment by NiLuJe** [comment]
> That's what I would expect, but note that I cleverly wrote "assuming" in my initial reply. ;-)

The code is fairly fresh in my mind given the base PR I linked to, so i'm confident making that one an assertion at this time ;).

**Comment by Frenzie** [comment]
Well, I don't recall it doing this in the past and I'm somewhat sure I tried (I mean a very long time ago, like 8 years) but it does seem to expand outside of the viewport.

The first picture shows the viewport. The second picture shows things outside of it. The third picture is a screenshot of the viewport.
![2024-01-03 20 43 19](https://github.com/koreader/koreader/assets/202757/c4630cbe-bedc-4f8a-855d-e56f26bf1548)
![2024-01-03 20 43 30](https://github.com/koreader/koreader/assets/202757/c11348c7-0909-4df2-a3da-a2d721553401)
![Reader_2024-01-03_204339](https://github.com/koreader/koreader/assets/202757/bf2b8700-1a73-4ef7-b858-f3536d144571)

**Comment by Frenzie** [comment]
SDL with `EMULATE_READER_VIEWPORT='{x=100, y=100, w=1095, h=1064}' ./kodev run -s=kobo-aura-one`
<img src=https://github.com/koreader/koreader/assets/202757/f85a9c83-925d-4dd3-ae6f-18955a63ca05 width=45%> <img src=https://github.com/koreader/koreader/assets/202757/2ef66e31-f120-4db4-acc4-0143667586a3 width=45%>

**Comment by Frenzie** [comment]
> There _might_ be an off-by-one somewhere in the rect rotation code (`bb:getPhysicalRect`), or the viewport offset accounting during refreshes (renamed to `fb:calculateRealCoordinates` in [koreader/koreader-base#1718](https://github.com/koreader/koreader-base/pull/1718)).

Both of these are without this PR though.

**Comment by poire-z** [comment]
(Didn't know about EMULATE_READER_VIEWPORT.)
Just tried it, and there's indeed something fishy: the remnants of thick line on the side do not even align with the ones show by the menu. (The end with nightmode and margin is me just being sure my crengine blue background take the full viewport width.)
![viewport-overflow](https://github.com/koreader/koreader/assets/24273478/ef7be834-79fe-4b2a-a528-3bd6b299a193)

**Comment by Frenzie** [comment]
@poire-z Yes, though the bothersome thing is that it behaves the same on a real Kobo device. At first I figured I must've simply messed up something fairly trivial on the SDL side of things.

**Comment by NiLuJe** [comment]
The refresh part, at least, is because ReaderMenu passes negative coordinates to `setDirty`, and the viewport offsetcomputation *honors* negative coordinates, instead of handling them sanely like the non-viewport codepaths.

I'll check whether something similar doesn't happen for blitting... somewhere.

**Comment by NiLuJe** [comment]
And the blitting is not a bug, it's an optimization: `paintRect` (and probably a lot of other things) can paint scanline-by-scanline, and to do so, it honors the buffer's *physical* stride (i.e., that includes off-screen scratch space/alignment spacing).

That explains why there's a one pixel offset on the *left* side, but not the right.

i.e., in the example below, black is the backing buffer, white is the viewport, and red is a theoretical single pixel (emphasized for effect ;p) as handled by paintRect.

![paintRect_stride](https://github.com/koreader/koreader/assets/111974/b35b3067-0e86-47da-9b0e-da81507adc5d)


TL;DR: Just need to fix the refresh.

**Comment by NiLuJe** [comment]
i.e., something like that that traced branch:

```diff
diff --git a/ffi/blitbuffer.lua b/ffi/blitbuffer.lua
index 4f9da335..20da79ac 100644
--- a/ffi/blitbuffer.lua
+++ b/ffi/blitbuffer.lua
@@ -1199,6 +1199,7 @@ function BBRGB32_mt.__index:blitToRGB32(dest, dest_x, dest_y, offs_x, offs_y, wi
 end
 
 function BB_mt.__index:blitFrom(source, dest_x, dest_y, offs_x, offs_y, width, height, setter, set_param)
+    print("BB:blitFrom", source, dest_x, dest_y, offs_x, offs_y, width, height, setter, set_param)
     width, height = width or source:getWidth(), height or source:getHeight()
     -- NOTE: If we convince CRe to render to a padded buffer (to match pixel_stride and allow us single-copy blitting),
     --       change the self:get* calls to self:getPhysical* ones ;).
@@ -1572,9 +1573,14 @@ paint a rectangle onto this buffer
 @param setter function used to set pixels (defaults to normal setPixel)
 --]]
 function BB_mt.__index:paintRect(x, y, w, h, value, setter)
+    print("paintRect", x, y, w, h, value, setter)
     setter = setter or self.setPixel
     value = value or Color8(0)
     x, y, w, h = self:getBoundedRect(x, y, w, h)
+    print("bounded paintRect:", x, y, w, h)
+    if x == 0 and w == self.w then
+        print(debug.traceback())
+    end
     if w <= 0 or h <= 0 then return end
     if self:canUseCbb() and setter == self.setPixel then
         cblitbuffer.BB_fill_rect(ffi.cast(P_BlitBuffer, self),
diff --git a/ffi/framebuffer_SDL2_0.lua b/ffi/framebuffer_SDL2_0.lua
index fca21859..7ea506cf 100644
--- a/ffi/framebuffer_SDL2_0.lua
+++ b/ffi/framebuffer_SDL2_0.lua
@@ -139,14 +139,10 @@ function framebuffer:refreshFullImp(x, y, w, h)
     if self.dummy then return end
 
     local bb = self.full_bb or self.bb
-
-    if not (x and y and w and h) then
-        x = 0
-        y = 0
-        w = bb:getWidth()
-        h = bb:getHeight()
-    end
-
+    print(string.format("refreshFullImp: %dx%d+%d+%d", w, h, x, y))
+    x, y, w, h = bb:getBoundedRect(x, y, w, h)
+    print(string.format("getBounded: %dx%d+%d+%d", w, h, x, y))
+    x, y, w, h = bb:getPhysicalRect(x, y, w, h)
     self.debug("refresh on physical rectangle", x, y, w, h)
 
     local flash = os.getenv("EMULATE_READER_FLASH")
```

Note to self: double-check the sanitization on that SDL refresh imp...

**Comment by NiLuJe** [comment]
This should be fixed in https://github.com/koreader/koreader-base/pull/1718 with the latest commits ;).

(Also fixed a mildly-related SDL-specific issue with rotation and the `EMULATE_READER_FLASH` codepath).

**Comment by NiLuJe** [comment]
Oh, and, FWIW, I was testing w/ `EMULATE_READER_FLASH=100 EMULATE_READER_VIEWPORT='{x=50, y=50, w=658, h=924}' ./kodev run -s=legacy-paperwhite` because that's the only thing that fit @ 1440p, and there's quite likely an issue somewhere when viewport and resizes and/or forced resizes are involved, because SDL crops a window because it doesn't fit on screen and there's a viewport involved, the viewport input translation is all kinds of fucked up.

Not going to bother with that one for now since platforms with a viewport don't support resizing windows anyway.

**Comment by Frenzie** [comment]
> Not going to bother with that one for now since platforms with a viewport don't support resizing windows anyway.

I don't think that's quite true, unless viewport support was removed from the Android implementation. That's why I once added support for dynamically updating the viewport at runtime.

**Comment by NiLuJe** [comment]
I suspect actual resizes might work, as opposed to whatever happens with SDL on startup.

No time to look into in the next few days though ;).
